### DISCUSSION TITLE
Beta UTF-8 region setting with Chinese language displays DnaLibrary AutoOpen "Memory" warning

### DISCUSSION THREAD

**Discussion by eutiquiorg** [Excel-DNA/ExcelDna#586] [Q&A]
Hello there! 
We have an add-in added to Excel and when we open Excel under the conditions described below (Steps to reproduce), a "Memory" warning is displayed and then we cannot use the available functions of our add-in.

We create a simple ClassLibrary project to reproduce our problem in Visual Studio 2022 (having Visual Studio in Chinese and having the "Steps to reproduce").
![image](https://user-images.githubusercontent.com/22947364/229930788-c0f446ca-983f-4b12-a427-a7de57946925.png)

Class1.cs
'namespace MyClassProject
{

    [ComVisible(true)]
    public class MyComAddIn : ExcelComAddIn
    {

        public MyComAddIn()
        {
        }

        public override void OnConnection(object Application, ext_ConnectMode ConnectMode, object AddInInst, ref Array custom)
        {
            SWF.MessageBox.Show("OnConnection");
        }

        public override void OnDisconnection(ext_DisconnectMode RemoveMode, ref Array custom)
        {
            SWF.MessageBox.Show("OnDisconnection");
        }

        public override void OnAddInsUpdate(ref Array custom)
        {
            SWF.MessageBox.Show("OnAddInsUpdate");
        }

        public override void OnStartupComplete(ref Array custom)
        {
            SWF.MessageBox.Show("OnStartupComplete");
        }

        public override void OnBeginShutdown(ref Array custom)
        {
            SWF.MessageBox.Show("OnBeginShutDown");
        }
    }

    public class MyAddIn : IExcelAddIn
    {
        private ExcelComAddIn _comAddIn;

        public void AutoOpen()
        {
            try
            {
                _comAddIn = new MyComAddIn();
                ExcelComAddInHelper.LoadComAddIn(_comAddIn);
            }
            catch (Exception e)
            {
                SWF.MessageBox.Show("Error loading COM AddIn: " + e.ToString());
            }
        }

        public void AutoClose()
        {
        }
    }
}`

MyClassProject.csproj
![image](https://user-images.githubusercontent.com/22947364/229909513-a1225afc-980e-4cca-bfe6-b48c56db902f.png)

We debug the code and when the debugger reaches ExcelComAddInHelper.LoadComAddIn(_comAddIn); the "Memory" warning is displayed as below picture.
![warning](https://user-images.githubusercontent.com/22947364/229904208-fc21527a-9e19-46c1-9989-dd3c8f661995.png)

Click on Continue and the following Exception is raised.
![image](https://user-images.githubusercontent.com/22947364/229933177-e51c17e8-4082-4e9e-8afd-26c33569397b.png)

This behavior occurs when we enable "Beta: Use Unicode UTF-8 for worldwide language support" option in the Region Settings and set the Windows language and Office language to Chinese (See "Steps to reproduce"). 

**Steps to reproduce:**
1.	Set the Regional format to "Chinese (Simplified, China)", as below picture
![image-20221207-134326](https://user-images.githubusercontent.com/22947364/229904844-e3190e26-31a8-4150-87ce-b57d0dd5a98d.png)
 
2.	Set the Region setting to use "Beta: Use Unicode UTF-8 for worldwide language support", and the Current system locale option as "Chinese (Simplified, China), as below picture
![image-20221207-134347](https://user-images.githubusercontent.com/22947364/229904909-335d81eb-6b04-4b12-85cf-69d1c99a790b.png)
 
3.	Set Excel as Chinese language in the File-> Options->Language -> Office display language as Chinese (Simplified), as below picture
![image-20221207-134404](https://user-images.githubusercontent.com/22947364/229905050-127e3358-b943-4a9e-9907-691f3bb0e182.png)
 
4.	Set Windows Display as Chinese
![image-20221207-134455](https://user-images.githubusercontent.com/22947364/229905084-1537bce6-b0c0-423c-ac04-bec773b5fad9.png)
 
5.	Restart Windows 10 system 
6.	Run the simple ClassLibrary project, you will see the "Memory" warning.

We appreciate your help with this issue.
Thank you!

**Comment by govert** [comment]
Wow - that is quite a story, and a very detailed write-up of the problem. I can't imagine how you found it and got to these details.

Certainly "Send a Frown" inside the Excel Feedback page (you can point here for the issues details) - I think the guys at Microsoft who are responsible for the Office add-in work are in Beijing, so if your issue gets routed there you have some hope of attention.

I imagine something is going wrong with the way COM add-ins are registered and then loaded of-the-fly by Excel-DNA. During that `ExcelComAddInHelper.LoadComAddIn` call we register the COM classes and the add-in itself in the registry (with the registry possibly being redirected by the AppV system that Office uses for Click-to-Run installations) then we call Excel via COM interop to load the add-in, which in turn reads the registry and makes the calls for the loading. Our writes to the registry go through the .NET registry classes.
So somewhere along this chain of .NET, Windows Registry and Excel, there might be a use of the ANSI code page which is set when you pick the UTF8 option. And in particular then interacting with the display language (which might historically have interacted with the Windows code page?).

So there might be a bug in the .NET registry classes, in the Windows registry support, in the Office registry redirection, in the Windows COM APIs or in Excel's loading of the add-in. And if I set the Chinese system and display language, I probably can't drive my computer or debugger very far.

Maybe one might start with some Process Monitor logging to track the registry reads and writes.
You can also make a simple COM add-in (no Excel-DNA involved and registered 'normally') and try to reproduce.

**Reply by eutiquiorg** [reply]
Thanks Govert for the detailed information and recommendations, we are following some of them to see if we can find more clues of the problem. We were able to identify that the "Memory" popup is launched when we get to this part [image 2] with these "XlCall" values (33122 and 6) [image 1]. 
![image](https://user-images.githubusercontent.com/22947364/231910502-58d14346-f310-4f88-8fd4-a56edde46e08.png)


Do you have any idea what is happening beyond this point? 
The debugger does not go beyond this line [image 2].
![image](https://user-images.githubusercontent.com/22947364/231910520-e05e93a6-2280-4402-af96-30d694bd03b7.png)
